### ISSUE TITLE
How to add new dependencies?

### ISSUE THREAD

**Issue by nkiesel** [jmfayard/buildSrcVersions#17]
_Context: [buildSrcVersions](https://github.com/jmfayard/buildSrcVersions) is a Gradle plugin that makes it easier to manage your dependencies inside your IDE. It extracts all your dependencies and search for available dependencies updates._

The description shows how to bootstrap the switch and how to update to new dependencies.  However, I can't see how to add a new dependency.  Looks to me that `Libs.kt` is not meant to be manually edited.  So is the idea to add an "old-fashioned" dependency, and then run syncLibs to create a new `Libs.kt`, and then replace the dependency literal?  That feels convoluted.  Is there a better way?  Perhaps a `registerDependency` task that takes a dependency string and adds it to `Libs.kt` and `Versions.kt`?

**Comment by jmfayard** [comment]
@nkiesel 

> However, I can't see how to add a new dependency. Looks to me that Libs.kt is not meant to be manually edited. That feels convoluted. I

That's correct. 

> Is there a better way?

I would suggest to use the IDE plugin [**OK, Gradle!**](https://github.com/scana/ok-gradle) from @scana  that solves the adding dependency problem brillantly. 

![Take a look](https://github.com/scana/ok-gradle/blob/master/images/look_and_feel.gif?raw=true)

Currently, you would still have to run `./gradlew syncTasks`  and replace the strings with `Libs.whatever`after. Ideally, I would like to integrate the IDE plugin with my gradle plugin. 

Related: https://github.com/scana/ok-gradle/issues/14

**Comment by jmfayard** [comment]
> Perhaps a registerDependency task that takes a dependency string and adds it to Libs.kt and Versions.kt?

You are right that we will also want to copy/paste dependencies from a README like this

```groovy
dependencies {
    def room_version = "2.1.0-alpha02"
    implementation "androidx.room:room-runtime:$room_version"
    kapt "androidx.room:room-compiler:$room_version" 
    testImplementation "androidx.room:room-testing:$room_version"
}
```

and to do something like this

```kotlin
// pipe the clipboard content to the "addLibs" task
$ pbpaste | ./gradlew addLibs 
Added 3 dependencies to buildSrc/src/main/kotlin/Libs.kt

dependencies {
    implementation(Libs.room_runtime)
    kapt(Libs.room_compiler)
    testImplementation(Libs.room_testing)
}    
```

Not exactly trivial but probably doable and super useful

**Comment by nkiesel** [comment]
> Not exactly trivial but probably doable and super useful

tl;dr: I would think that automating the "register in Libs/Versions" would be enough and we can leave adding the dependency as a manual step.

Yes, that would be the "gold-plated" version. However, also the most complicated version, especially the "automatically add to build.gradle.kts" part because of the flexible/creative way that different people structure that file.  The part of adding them to "Versions/Libs" seems much simpler (compared with that) because they are under your control.  And once that is done, manually adding the actual dependencies is already very pleasant with the IDE completion for Libs.*

Another concrete point: in one big project we have a multi-module setup with a top-level gradle file that contains all the versions (in a hand-written groovy array), and modules-specific build.gradle then add dependencies they need.  I would love to switch that over to your wonderful syncLibs (after switching to Kotlin DSL first) but in this setup automatically adding dependencies would never be useful (or really complicated: when called inside a module, register in the top-level buildSrc and add to the module-specific build.gradle.kts).

**Comment by jmfayard** [comment]
@nkiesel 
> tl;dr: I would think that automating the "register in Libs/Versions" would be enough and we can leave adding the dependency as a manual step.

Makes sense!

> Another concrete point: in one big project we have a multi-module setup with a top-level gradle file that contains all the versions (in a hand-written groovy array), and modules-specific build.gradle then add dependencies they need. I would love to switch that over to your wonderful syncLibs (after switching to Kotlin DSL first)

I'm not sure to understand, is there something that prevents you from switching?
Applying `syncLibs` to a such a project works. It generates the right Libs.kt and Versions.kt, then you use Libs.*, then as a last step, you can remove the groovy hashmap that is now un-used.

You don't need to switch to the Kotlin DSL first, it works the same with a Groovy build. 
In fact, if you plan to switch to the Kotlin DSL, I would recommend to leverage `syncLibs` first, as I did here https://github.com/rozkminiacz/KotlinUnitTesting/pull/2

> but in this setup automatically adding dependencies would never be useful (or really complicated: when called inside a module, register in the top-level buildSrc and add to the module-specific build.gradle.kts).

Right, it's not in my intention to do that either.

**Comment by nkiesel** [comment]
I somehow had it in my head that syncLibs would require the switch to Kotlin DSL, so thanks for pointing out that this is not necessary!  I will give it a shot.

**Comment by nkiesel** [comment]
Does this "moved this from To do to Done" indicate that this idea is now implemented?  I looked a bit around in the 0.4.1 version but could not find any changes related to this.

**Comment by jmfayard** [comment]
No new change in the plugin, I improved the documentation last year

**Comment by jmfayard** [comment]
@nkiesel there are finally news on this topic.
You can do `$ ./gradlew :buildSrcVersions --update" or `$ ./gradlew refreshVersions --update`.
It will upgrade to the newer versions and you can fix what you don't want by looking at the git diff.
This will makes it permanent

```kotlin
buildSrcVersions {
    alwaysUpdateVersions()
}
```

https://github.com/jmfayard/buildSrcVersions/issues/53
